### DISCUSSION TITLE
Post v1.0.0 new features

### DISCUSSION THREAD

**Discussion by e-t-k** [CBIIT/ccdi-federation-api#110] [API Design Discussion]
This discussion is to track new features in the  **interface design** that we would like to add post v1.0.0 release. This can be large features that will be added in v2.0.0, or smaller features that will be added in a v1.x patch.

This is not for bugs (see #109 for that)  and it is not for any particular implementation's feature completion status.

#### ADDED Add Two Permissible Values to the File Type element
~**Endpoint** GET /file
**Desired result** File Type matches the [File Type CDE in caDSR ](https://cadsr.cancer.gov/onedata/dmdirect/NIH/NCI/CO/CDEDD?filter=CDEDD.ITEM_ID=11416926%20and%20ver_nr=1) with 70 permissible values, including gVCF and HTSeq Count
**Current version** gVCF and HTSeq Count are not present as permissible values for file type
**Is this a breaking change?** Yes: response for File.type may now include new values which clients might not expect or be able to process.
**Will implementations need to update?** No: If implementation has no files with type gVCF or HTSeq Count this change does not affect them.
**Date added** June 24 2024
**Current status** Completed Dec 17 2024 in https://github.com/CBIIT/ccdi-federation-api/commit/f56324548a63e3bff0305a1db244ffb3efc51fb4
**Notes** These permissible values were added to caDSR by request of the Federation ([gVCF](https://github.com/CBIIT/ccdi-federation-api/discussions/53#discussioncomment-8248652), [HTSeq Count](https://github.com/CBIIT/ccdi-federation-api/discussions/53#discussioncomment-8973686)) subsequent to the File Type element being added to the API spec.~

**Comment by Ssandor13** [comment]
**New Features**

**Fix Missing Permissible Values and Remove 2 Permissible Values to the [Disease Phase](https://github.com/CBIIT/ccdi-federation-api/wiki/Sample-Metadata-Fields#disease_phase) element**

**Endpoint** GET /sample
**Desired result** caDSR Link: [12217251](https://cadsr.cancer.gov/onedata/dmdirect/NIH/NCI/CO/CDEDD?filter=CDEDD.ITEM_ID=12217251%20and%20ver_nr=1) 
**Current version** does not show `Relapse/Refractory` `Remission` `Tumor Surveillance` but shows `Unknown` and `Not Reported` 
**Is this a breaking change?** Yes: response may now include new values which clients might not expect or be able to process.
**Will implementations need to update?** No: If implementation has no samples with types aforementioned above, this change does not affect them.
**Date added** July 8th, 2024
**Current status** feature not yet started as of July 8th, 2024

**Reply by GeoffLyle** [reply]
This is odd that the CDE would be changed after being released.  @jknable Should we expect CDEs to change and continue to use the same version number?

**Reply by jknable** [reply]
The permissible values of CDE 12217251 (Disease Phase Timepoint Category) were last updated on 12/26/23 and 12/28/23 which removed Unknown and Not Reported and added 3 new permissible values (Relapsed/Refractory, Remission, Tumor Surveillance).  Since the CDE had a status of “Draft New” at that time, no versioning was required.  

CDE versioning rules require that once released, a CDE is versioned when a permissible value is removed, but not when new permissible value is added.  There is a unique “Begin Date” attribute associated with each permissible value which can be used to identify updates.

**Comment by Ssandor13** [comment]
**New Features**

**Remove 10 Permissible Values [Tissue Preservation](https://github.com/CBIIT/ccdi-federation-api/wiki/Sample-Metadata-Fields#preservation_method) element**

Endpoint GET /sample
Desired result caDSR Link: [5432521](https://cadsr.cancer.gov/onedata/dmdirect/NIH/NCI/CO/CDEDD?filter=CDEDD.ITEM_ID=5432521%20and%20ver_nr=1.0)
Current version shows too many permissible values and does not match caDSR Link: [5432521](https://cadsr.cancer.gov/onedata/dmdirect/NIH/NCI/CO/CDEDD?filter=CDEDD.ITEM_ID=5432521%20and%20ver_nr=1.0)
Will implementations need to update? Maybe: If implementation has no samples with types aforementioned in caDSR link, this change does not affect them, however, if the Federation used the wiki to map to then the implementation would need to be updated.
Date added July 8th, 2024
Current status feature not yet started as of July 8th, 2024

**Reply by GeoffLyle** [reply]
From the discussion on Preservation Method #64, we went with CDE [8028962](https://cadsr.cancer.gov/onedata/dmdirect/NIH/NCI/CO/CDEDD?filter=CDEDD.ITEM_ID=8028962%20and%20ver_nr=2) since it contains the **EDTA** permissible value.

**Reply by Ssandor13** [reply]
@GeoffLyle thank you. I apologize as I went with the link at the top of the discussion rather than reading through the entire discussion; I assumed that was the one we were using given it was at the top.

**Comment by asafieva** [comment]
CDE 6273393 v1.00 sample>>library_strategy 
Two PVs were not included in wiki and in swagger.yaml "cde.v1.sample.LibraryStrategy": scMultiome, Spatial-tx. 
Was this feature implemented as expected and Federation API needs only a subset of this CDE PVs? These PVs are not returned by any participating node implementations to the moment.
**Date added** July 18th, 2024
**Current status** feature not yet started as of July 18th, 2024

**Reply by asafieva** [reply]
Not all CDE PVs are included into Harmonized API list of values.

**Comment by Ssandor13** [comment]
@jknable `Ethnicity` was showing a lower case `r` instead of capital `R` for `Not Reported` permissible value. See snapshot below. There may be others with this issue as well. This will not only affect our wiki, mappings per Federation, but the API spec as well. 
![ethnicity_not_reported (2)](https://github.com/user-attachments/assets/a65aba81-fbdf-4ee3-9173-31546334d8b4)

**Date added:** August 2nd, 2024
**Current status:** feature not yet started as of August 2nd, 2024.

**Reply by e-t-k** [reply]
As of Feb 19 2025, the permissible values for this look correct in the [reference implementation](https://github.com/CBIIT/ccdi-federation-api/blob/main/crates/ccdi-cde/src/v2/subject/ethnicity.rs) -- I'm seeing lowercase `Not reported` which matches the Permissible Value column of the CDE screenshot.

**Comment by GeoffLyle** [comment]
**New Features**

**Add 2 new CDEs accepted by the Data Harmonization team.**

Library Source Material ~~#114~~ #119 - ~~CDE [14808227](https://cadsr.cancer.gov/onedata/dmdirect/NIH/NCI/CO/CDEDD?filter=CDEDD.ITEM_ID=14808227%20and%20ver_nr=1)~~ CDE [15235975](
https://cadsr.cancer.gov/onedata/dmdirect/NIH/NCI/CO/CDEDD?filter=CDEDD.ITEM_ID=15235975%20and%20ver_nr=1)
Library Selection Method #47 - CDE [6347743 v2](https://cadsr.cancer.gov/onedata/dmdirect/NIH/NCI/CO/CDEDD?filter=CDEDD.ITEM_ID=6347743%20and%20ver_nr=2)

Note: For Library Selection Method we agreed to use a subset of the values in the CDE.  These values are:

- Hybrid Selection
- Not applicable
- ~~PCR~~
- Poly-A Enriched Genomic Library
- Random
- Random PCR
- rRNA Depletion
- Unspecified

[Update](https://github.com/CBIIT/ccdi-federation-api/discussions/47#discussioncomment-11007066) on October 21, 2024.

**Reply by GeoffLyle** [reply]
Note: Library Source Material https://github.com/CBIIT/ccdi-federation-api/discussions/114 has been reopened for Data Harmonization team discussion on October 04, 2024.

**Reply by GeoffLyle** [reply]
Note: After discussion on October 18, 2024, Library Source Material https://github.com/CBIIT/ccdi-federation-api/discussions/114 has been re-approved for implementation.  Federation members will map blood tumor samples to "Bulk Tissue".

In the future the Federation would like to add a CDE that differentiates among "blood", "saliva", "tissue", "tears". (see https://github.com/CBIIT/ccdi-federation-api/discussions/108#discussioncomment-9863505)

**Reply by GeoffLyle** [reply]
Note: After discussion on November 1, 2024, the Federation will be implementing the CCDI CDE for Library Source Material (CDE [15235975](https://cadsr.cancer.gov/onedata/dmdirect/NIH/NCI/CO/CDEDD?filter=CDEDD.ITEM_ID=15235975%20and%20ver_nr=1)).  
This CDE contains the same Permissible Values as the previous CDE, but has an updated definition "The cell or cellular component that makes up the sample that has been prepared for testing or research purposes." which will allow for more flexibility in the future.

**Reply by GeoffLyle** [reply]
Note: Library Source Material and Library Selection Method can be implemented as part of the /sample endpoint.

**Reply by GeoffLyle** [reply]
Library Source Material and Library Selection Method have been implemented in the /sample endpoint.  This information has been added to the GitHub Wiki as well as the Swagger document.  UCSC, CHOP, St. Jude (members with the /sample endpoint) are implementing both of these fields.

**Comment by asafieva** [comment]
This is a question about sharing "**_study_**" information through CCDI API.
The only place to receive _**"study"**_ information in API v1.0.0 is using the "[namespace](https://federation.ccdi.cancer.gov/api/v1/namespace)" endpoint. The component containing study attributes is "models.namespace.Metadata".
Besides KidsFirst, all other APIs nodes send "null" as namespace metadata, means no study information is sent.
Is there any plan to include "study" information into "namespace" endpoint API responses? Is there any other feasible way to connect "study" information to API entities such as "subject"? 
[CCDI hub](https://ccdi.cancer.gov/explore) and [CPI](https://participantindex-docs.ccdi.cancer.gov/) extensively use study references. 

Here is a consideration given by Brian Furner (PCDC) in the slack API channel.
_It feels like information about studies could be a first order entity in a data model and right now they’re only represented as metadata about a namespace which is highly abstract.
In general, PCDC captures enrollment information for individual persons on one or more COG protocols, but what we have is just the study_id. It feels like we may want a discussion on how to handle reference data here as it is possible if not likely that there will be the same studies represented in different member source systems and unless we coordinate on this, search will be a challenge._

**Comment by GeoffLyle** [comment]
**New Features

Add 1 new CDEs accepted by the Data Harmonization team (October 4, 2024).**

Specimen Molecular Analyte Type #116 - CDE [15063661](https://cadsr.cancer.gov/onedata/dmdirect/NIH/NCI/CO/CDEDD?filter=CDEDD.ITEM_ID=15063661%20and%20ver_nr=1)

Permissible values have been subset to:

- DNA
- Protein
- RNA

Note: Values 'Not Reported' and 'Unknown' are not currently included.  This information should be known when sharing genomic data. If a member does not fill out this field, the API should return a _null_ response. We believe not including these values will make this field easier for users to parse.

**Reply by GeoffLyle** [reply]
Note: Specimen Molcular Analyte Type can be implemented as part of the /sample endpoint.

**Reply by GeoffLyle** [reply]
Specimen Molecular Analyte Type has been implemented in the /sample endpoint. This information has been added to the GitHub Wiki as well as the Swagger document. UCSC, CHOP, St. Jude (members with the /sample endpoint) are implementing both of these fields.
